### ISSUE TITLE
Generated Java cannot handle Optional Fields properly

### ISSUE THREAD

**Issue by texastony** [awslabs/smithy-dafny#278]
#### Issue
A Smithy Structure with an Optional Number member is represented in Dafny Source as
`Option<int32>`.

In Smithy Generated Java, the type for this member on the Structure's class will be `int`. 

`null` cannot be assigned to a Java `int`. 

As such, 
`Wrappers_Compile.Option.None` cannot be represented by 
the Java type.

#### Suggested Fix V2.1
To avoid a breaking change AND resolve this issue,
the only fix I can think of is to follow the pattern established by .NET implementation.

The "set-state" of a member MUST be tracked.
The "set-state" is a boolean flag that defaults to `false`.
The "set-state" switches to `true` when a value is provided for the member
during a Structure Shape's construction.
When a Structure is validated,
if the "set-state" of a  `@required` member is `false`,
the validation MUST fail.

In Java, 
this means the Builders of Structures
needs to be refactored to generate and utilize
"set-state" flags for each member.

The "set-state" flags need to be included on the final "built" structure.

In the ToDafny and ToNative classes,
these "set-state" flags can then be used to determine if no value was set.

#### Breaking Change Fix that is better.
Avoid Java Primitive types all together.
Or at least for non-required members.

This is going to complicate the type conversion.

Dafny Generated Java represents Optional Members as `Wrappers_Compile.Option<primitive>`.
Thus, we will be converting from non-primitive types (`Integer`) to primitive types (`int`).

An alternative approach would be to use Java's `Optional`.

However, Java discourages using `Optional` as a parameter or field;
it SHOULD BE exclusively used a Return type.

**Comment by robin-aws** [comment]
The ideal pattern would be to follow what the current Java v2 SDKs do: required members use primitives like `int` directly, optional members use `Integer` so they can be `null`. E.g. https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/sqs/model/SendMessageRequest.html#delaySeconds()

Edit: it's possible even required members use `Integer` as well, I didn't immediately find one I know is required to confirm.

**Comment by texastony** [comment]
I searched around, and could not find AWS SDK V2 primitives. 

From a "parsing customer input" perspective, I can see the benefit of having nullable fields. 

It affords differentiation between a customer setting/providing a value and not. 

I'm surprised I cannot find an opinion on this in the V2 Smithy Spec. 

I found it in the V1 Spec.

**Comment by robin-aws** [comment]
> Suggested Fix V2
To avoid a breaking change AND resolve this issue,
the only fix I can think of is to follow the pattern established by .NET implementation.
...

I think this is the right direction, and oddly enough we already include boolean flags like `_GreaterThanOneSet` on output structures but not input structures (Edit: I was confused, we define these boolean flags on the builders for inputs and outputs, but not the built structures themselves). We also only use these flags to ensure validation isn't applied to non-existent values, but the behavior will still be surprising because there's no public accessor to this state, so a `GreaterThanOne()` result of `0` is indistinguishable from no value set, and yet `0` clearly violates `range(min: 1)`.

For the record the original suggested fix of using reference types like `Integer` in the builder/structure interfaces is definitely the better user experience and what the Java SDKs do. The above fix will mean user code will have to do this:

```java
if (output.CountIsSet()) {
  ProcessCount(output.Count());
}
```

As opposed to if Count() returned an `Integer` instead of an `int`:

```java
if (output.Count() != null) {
  ProcessCount(output.Count());
}
```

Given that I'm inclined to implement both versions and pick which via an option. The latter can be the default in a new edition value for `smithy-dafny` so that backwards compatibility of existing libraries is maintained.

> Alternatively, we could move input Validation into the Generated Dafny

I fully agree and work is already underway to make this possible! In the meantime it makes improving testing coverage here even more important, so that we can safely make that move in the future.

**Comment by texastony** [comment]
"Perfect is the enemy of Done."

We could get the "set-state" flag fix out sooner,
and leave this issue open,
or cut a new one,
to track the better but breaking fix.

I say this because the better-but-breaking fix MAY not be used by Crypto Tools for YEARS.
We try to not make new Major Versions of products vary often...

**Comment by robin-aws** [comment]
Yeah agreed that fixing the current API choice is more urgent. I'm talking about also quickly improving things so new smithy-dafny projects don't also pick up the suboptimal UX.

I do expect Crypto Tools will probably never switch now that there are three libraries out there with this version. But I don't want smithy-dafny to be stuck with that version permanently.

Note I mention a new `Edition` for smithy-dafny, so that Crypto Tools projects can still pick up the latest smithy-dafny while specifying an option like `--with-is-set-methods-in-java` or whatever to maintain current APIs for the forseeable future. No major version bump of smithy-dafny, and no change at all to the generated code in the Crypto Tools projects.

**Comment by robin-aws** [comment]
I should also say using `Integer` is safer, because you can't forget to call `CountIsSet()` before calling `Count()` and getting an inaccurate result.

**Comment by robin-aws** [comment]
Just to record offline decisions on this one:
1. We've decided just replacing `int` with `Integer` is the better fix: even though it is technically breaking, we suspect it will not break anyone in practice, and this risk is worth it to provide a simpler and safer interface. We expect it will be too easy to not be aware of separate `fooIsSet` methods and hence accidentally misinterpret values.
2. Along the same lines, we're going to make the same changes for .NET to replace (for e.g.) `int` with `int?`

**Comment by texastony** [comment]
For clarity, the changes in .NET will not remove the need the call `Validate` at type conversion.

Changing the .NET structures to use builders would be a breaking change.

**Comment by texastony** [comment]
Blocks https://github.com/aws/aws-cryptographic-material-providers-library/pull/80

**Comment by seebees** [comment]
This work is being completed in #496

**Comment by texastony** [comment]
FYI: I think the Java generator respects the `box` trait. 

So changing Smithy 1.0 models to use it would suffice for some instances. 

However, the Java generator has not, to my knowledge, been updated for Smithy 2.0.
